compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp

```diff
@@ -183,6 +183,57 @@ struct ConvertTensorFromElementsPattern
   }
 };
 
+/// Populates given sizes array from type (for static sizes) and from
+/// the tensor (for dynamic sizes).
+static void sizesForTensor(OpBuilder &builder, SmallVectorImpl<Value> &sizes,
+                           Location loc, ShapedType stp, Value tensor) {
+  for (const auto &d : enumerate(stp.getShape())) {
+    Value dim;
+    if (d.value() == ShapedType::kDynamic)
+      dim = builder.create<tensor::DimOp>(loc, tensor, d.index());
+    else
+      dim = builder.create<arith::ConstantIndexOp>(loc, d.value());
+    sizes.push_back(dim);
+  }
+}
+
+/// Convert tensor.reshape ops into flow.tensor.reshape ops where possible.
+struct ConvertTensorReshapeOpPattern
+    : public OpRewritePattern<tensor::ReshapeOp> {
+  using OpRewritePattern<tensor::ReshapeOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(tensor::ReshapeOp op,
+                                PatternRewriter &rewriter) const override {
+      if (op->getParentOfType<Flow::DispatchWorkgroupsOp>()) {
+        return failure();
+      }
+      auto loc = op.getLoc();
+      Value input = op.getSource();
+      ShapedType inputType = llvm::dyn_cast<ShapedType>(input.getType());
+      ShapedType resultType =
+          llvm::dyn_cast<ShapedType>(op.getResult().getType());
+      if (!inputType || !resultType || !inputType.hasRank() ||
+          !resultType.hasRank()) {
+        return rewriter.notifyMatchFailure(op, "not ranked shaped types");
+      }
+
+      SmallVector<Value> srcSizes;
+      sizesForTensor(rewriter, srcSizes, loc, inputType, input);
+
+      FailureOr<Value> output = tensor::getOrCreateDestination(rewriter, loc, op.getOperation()->getResult(0));
+      if (failed(output)) {
+        op.emitError() << "unable to get/create destination tensor";
+      }
+      Value outTensor = output.value();
+
+      SmallVector<Value> destSizes;
+      sizesForTensor(rewriter, destSizes, loc, resultType, outTensor);
```

Comment:
I don't think this will quite work. You are essentially getting the sizes of the result of the reshape, and then replacing that same result with the Flow reshape, so the result of the original reshape is still live. You need to get the tensor sizes from the 1D index tensor (2nd argument of tensor.reshape instead).

I would keep the unranked case unsupported for now and use a sequence of `tensor.extract` operations on the index tensor to get the target shape. In most cases this index tensor should come from a `tensor.from_elements` so I would expect most of that to fold away.

Reply:
Yeah, that's why I closed it actually. Wanted to make that change. It has been made now.